common/src/main/java/tachyon/security/authentication/PlainSaslServer.java

```diff
@@ -0,0 +1,255 @@
+      // authorizationId may be empty,then the authorizationId = authenticationId
+      String payload;
+      try {
+        payload = new String(response, "UTF-8");
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Received corrupt response", e);
+      }
+      String[] parts = payload.split("\u0000", 3);
+      // validate response
+      if (parts.length != 3) {
+        throw new IllegalArgumentException("Invalid message format, parts must contain 3 items");
+      }
+      String authorizationId = parts[0];
+      String authenticationId = parts[1];
+      String passwd = parts[2];
+      if (authenticationId == null || authenticationId.isEmpty()) {
+        throw new IllegalStateException("No authentication identity provided");
+      }
+      if (passwd == null || passwd.isEmpty()) {
+        throw new IllegalStateException("No password provided");
+      }
+      if (authorizationId == null || authorizationId.isEmpty()) {
+        authorizationId = authenticationId;
+      } else if (!authorizationId.equals(authenticationId)) {
+        // TODO: support impersonation
+        throw new UnsupportedOperationException("Impersonation is not supported now.");
+      }
+
+      NameCallback nameCallback = new NameCallback("User");
+      nameCallback.setName(authenticationId);
+      PasswordCallback passwordCallback = new PasswordCallback("Password", false);
+      passwordCallback.setPassword(passwd.toCharArray());
+      AuthorizeCallback authCallback = new AuthorizeCallback(authenticationId, authorizationId);
+
+      Callback[] cbList = {nameCallback, passwordCallback, authCallback};
+      mHandler.handle(cbList);
+      if (!authCallback.isAuthorized()) {
+        throw new SaslException("AuthorizeCallback authorized failure");
+      }
+      mAuthorizationId = authCallback.getAuthorizedID();
+
+      // After verification succeeds, a user with this authz id will be set to a Threadlocal.
+      AuthorizedClientUser.set(mAuthorizationId);
+    } catch (Exception e) {
+      throw new SaslException("Plain authentication failed: " + e.getMessage(), e);
+    }
+    mCompleted = true;
+    return null;
+  }
+
+  @Override
+  public boolean isComplete() {
+    return mCompleted;
+  }
+
+  @Override
+  public String getAuthorizationID() {
+    checkNotComplete();
+    return mAuthorizationId;
+  }
+
+  @Override
+  public byte[] unwrap(byte[] incoming, int offset, int len) {
+    throw new UnsupportedOperationException("PLAIN doesn't support wrap or unwrap operation");
+  }
+
+  @Override
+  public byte[] wrap(byte[] outgoing, int offset, int len) {
+    throw new UnsupportedOperationException("PLAIN doesn't support wrap or unwrap operation");
+  }
+
+  @Override
+  public Object getNegotiatedProperty(String propName) {
+    checkNotComplete();
+    return Sasl.QOP.equals(propName) ? "auth" : null;
+  }
+
+  @Override
+  public void dispose() {
+    if (mCompleted) {
```

Comment:
As far as I know, the threadLocal variable doesn't required. The PlainSaslServer will negotiate with client side when the first request was sent. The negotiate procedure will call evaluateResponse method. When the negotiation was done, the second request from the same thread doesn't need to negotiate with the SaslServer.